Adapt flex child controls for Spacer.

tellthemachines opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Fixes #38022.

* Removes custom size controls when Spacer is a child of a flex layout block (Row, Stack, Navigation).
* Adapts the block resizing handles so they change `flex-basis` value instead of width and height.

Note: I've made the "Fit" control reset the block width/height to 0 instead of removing it altogether, as removing "Fit" presents a technical challenge, due to those controls not knowing about the blocks they're applied to. I'm not sure if it makes sense to try and hack around this because Spacer is probably the only block that has no intrinsic size.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

1. In a post or template, add a Row block with a few blocks inside, including a Spacer;
2. Check that custom sizing controls no longer show in the Spacer sidebar;
3. Resize Spacer using its resizing handles and check that all works correctly;
4. Try setting Spacer to "Fill" with the child flex controls in the sidebar.

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

<img width="1029" alt="Screenshot 2023-03-27 at 5 08 44 pm" src="https://user-images.githubusercontent.com/8096000/227855019-56e5871a-f3b3-42f2-a9cf-62e507d2f455.png">

<img width="1030" alt="Screenshot 2023-03-27 at 5 08 59 pm" src="https://user-images.githubusercontent.com/8096000/227855040-ad528de6-d8bb-45e3-be6c-95da014dd4ca.png">

tellthemachines committed:
Adapt flex child controls for Spacer.

jasmussen commented:
Nice, thank you for tackling this! Already seems like a nice iteration. One thing I'm realizing is that the navigation container itself might also need a width control. This GIF shows a navigation block that's only as wide as its contents inside, and so when I set a "fill" property to a spacer inside, it's zero wide because there's no space for it to fill out:
![spacer](https://user-images.githubusercontent.com/1204802/227886807-929228a7-129b-4b37-b08b-22b44ff6e57c.gif)

<img width="464" alt="content wide container" src="https://user-images.githubusercontent.com/1204802/227886859-edeb42fd-84f6-46f2-9d58-7aee59aa1bb1.png">

Presumably the "fill" property would work if the navigation container was 100% wide. 

What do you think?

tellthemachines commented:
> This GIF shows a navigation block that's only as wide as its contents inside, and so when I set a "fill" property to a spacer inside, it's zero wide because there's no space for it to fill out

If that Nav block is inside a Row block, it'll have its own fit/fill/fixed controls under "Dimensions" (you might have to look for them in the dropdown because they don't show by default). If it's not inside Row then it should take up 100% of its container's width already.

andrewserong commented on packages/block-library/src/spacer/edit.js:
```diff
@@ -100,30 +105,80 @@ const SpacerEdit = ( {
+		if ( isFlexLayout && selfStretch === 'fit' ) {
+			return undefined;
+		}
+		return temporaryHeight || height || undefined;
```
Do we need to also use `style.layout.flexSize` in this value to sync changes to `flexSize` back to the styles applied to the Spacer?

andrewserong commented on packages/block-library/src/spacer/edit.js:
```diff
@@ -100,30 +105,80 @@ const SpacerEdit = ( {
+		if ( isFlexLayout && selfStretch === 'fit' ) {
+			return undefined;
+		}
+		return temporaryWidth || width || undefined;
```
Same comment as above: do we need to also use `style.layout.flexSize` in this value to sync changes to `flexSize` back to the styles applied to the Spacer?

andrewserong reviewed (commented):
Very cool idea, the child flex controls are going to be much nicer for playing around with spacing!

The syncing from dragging the handle on the spacer block to the Fixed width/height value in the flex controls appears to be working quite well. Unfortunately it looks like it isn't syncing the other way around when adjusting the value in either the horizontal or vertical orientations. I think it might be a matter of including the `flexSize` value in the `getHeightForVerticalBlocks` and `getWidthForHorizontalBlocks` functions?

https://user-images.githubusercontent.com/14988353/228102250-646084e0-9ab1-4e21-b11d-b385b14883d3.mp4

---

Other than that, now that the spacer controls are hidden for flex layouts, should we display the child flex controls by default so that the controls aren't hidden when a block is freshly inserted?

tellthemachines committed:
Resize from fixed input field

tellthemachines commented:
Thanks for reviewing! I've addressed the issue with resizing when the fixed value changes.

I also noticed something else: the resizing handles always convert whichever unit is used in the input field back to `px`. It's not great but it's existing behaviour so better fix it separately.

> Other than that, now that the spacer controls are hidden for flex layouts, should we display the child flex controls by default so that the controls aren't hidden when a block is freshly inserted?

I'm starting to feel like they should always display for all blocks because I've had to explain multiple times where to find them, which probably means they aren't discoverable enough.

andrewserong commented:
> I'm starting to feel like they should always display for all blocks because I've had to explain multiple times where to find them, which probably means they aren't discoverable enough.

Now that you mention it, I think that would make a lot of sense. Typically the controls that are hidden by default are ones that we think folks mightn't reach for all that frequently. When a block is a child of the Flex layout, there's a much higher likelihood that people are going to want to make adjustments to it. Might need some design feedback, and there could be some exceptions (Image block perhaps?) where there are other controls that might need to be more prominent than the flex children controls, but sounds like a good thing to explore to me!

jasmussen commented:
> I also noticed something else: the resizing handles always convert whichever unit is used in the input field back to px. It's not great but it's existing behaviour so better fix it separately.

That seems okay to me. How would the resize handle be able to work with, say, percentages? (Apologies if I'm missing something obvious.)

> I'm starting to feel like they should always display for all blocks because I've had to explain multiple times where to find them, which probably means they aren't discoverable enough.

Only on my second coffee so my mind isn't entirely clear. But at this juncture, I'm also _leaning_ that way. However

* Seems like we should do that separately
* I also think the ellipsis/show-hide functionality can be revisited so you shouldn't have to explain this

That is to say, I still think there's a place for hiding advanced functionality by default and adding it, but mainly that it's not always clear what is advanced vs. what is basic.

I also think @SaxonF may have had some thoughts around some of these dimensions migrating to the layout panel. I don't recall all the details, but just connecting the dots.

apeatling commented:
This seems to be working as expected in terms of the changes here. :+1:

Unrelated to this PR I noticed was that the tooltip shows "null" when resizing. I realize this is reflecting a height value that isn't set, but it felt like a bug and would be better to say something else? Is this related to the px conversion?

<img width="1240" alt="Screenshot 2023-03-28 at 10 44 14 AM" src="https://user-images.githubusercontent.com/1464705/228324574-7c978cd9-55f3-4253-9378-884da1cba1e5.png">

tellthemachines commented:
> How would the resize handle be able to work with, say, percentages? (Apologies if I'm missing something obvious.)

They might not work with percentages - that's probably why the custom Spacer unit control doesn't allow them - but we could conceivably make them work with rem, em, vw and vh. I agree it's not a super high priority though! 

> Unrelated to this PR I noticed was that the tooltip shows "null" when resizing

I'm not sure why this happens but given it's also happening on trunk I thought to investigate it separately.

tellthemachines commented:
I've solved the conflicts from the merge of #44002 - cc @glendaviesnz - and I think this is ready for another round of reviews!

andrewserong reviewed (commented):
This is generally working pretty well, but I noticed a couple of inconsistencies with how the flex child controls appear to work:

* In a Row or Stack block, when you insert a Spacer block, the flex child controls default to `Fit`, but the spacer visually looks like it uses its default value (`100px`). Should it be defaulting to the `Fixed` tab?
* The same issue occurs if you set a Spacer block outside of the flex block, and then drag the block inside the flex block. The tab is on `Fit` instead of `Fixed`. Also, if you toggle between the options, `Fit` always results in `0` size because there's no content.

Here's a quick screengrab of dragging a Spacer block into a Stack block:

![2023-04-12 11 18 24](https://user-images.githubusercontent.com/14988353/231322801-48b16e6f-a6a0-47d4-beb7-4b4d0313e653.gif)

Other than that, I also noticed that Fill in the vertical orientation sometimes behaves inconsistently for me if I don't have a min height value set. With min height set, it appears to be working quite well, though 👍

tellthemachines committed:
Make sure initial state is "Fixed"

tellthemachines commented:
Thanks for testing @andrewserong ! 

> In a Row or Stack block, when you insert a Spacer block, the flex child controls default to Fit, but the spacer visually looks like it uses its default value (100px). Should it be defaulting to the Fixed tab?

Yes, I've fixed that now.

>  if you toggle between the options, Fit always results in 0 size because there's no content.

That's expected given Fit is based on content size. Ideally, we wouldn't display Fit at all in this context, but I can't think of any good way of doing that doesn't involve adding a block attribute to manage child layout controls. Does that sound like a terrible idea? 😅 

> Other than that, I also noticed that Fill in the vertical orientation sometimes behaves inconsistently for me if I don't have a min height value set.

Can you provide more details on the inconsistency? Fill shouldn't really work unless the parent has a min height, but I notice it seems to default to 100px.

andrewserong commented on packages/block-library/src/spacer/edit.js:
```diff
@@ -197,6 +254,18 @@ const SpacerEdit = ( {
+					...blockStyle,
+					layout: {
+						...layout,
+						flexSize: width || '72px',
```
Does this need to account for `height` and/or the spacing presets values? When dragging into a Stack variation, it looks like it defaults to the `72px` value:

![2023-04-12 16 48 11](https://user-images.githubusercontent.com/14988353/231374667-757ad173-5589-49ab-8468-408431e95bc4.gif)

tellthemachines replied:
Ah yes, well spotted! I've fixed the logic in the effect now and the problem with fill state should be addressed.

andrewserong reviewed (commented):
Thanks for the update! Just left a comment about dragging a Spacer into the Stack block, it looks like the value might need to be updated there to also handle `height`.

For the Fill issue, I couldn't quite work out what's going on, but hopefully this video will help. It seems that the value set by dragging the control winds up overriding the Fill value possibly? Essentially, when I drag the spacer via the handle, then when I switch between Fixed and Fill, it'll remember the size based on the drag handle, rather than appearing to use Fill, if that makes sense? Here's a screengrab:

https://user-images.githubusercontent.com/14988353/231376091-64be4e7b-a540-480e-b3b2-6ebc0b98da09.mp4

Happy to give it another test tomorrow!

andrewserong commented:
>  Ideally, we wouldn't display Fit at all in this context, but I can't think of any good way of doing that doesn't involve adding a block attribute to manage child layout controls.

I suppose another possibility would be an attribute in `block.json` for child layout supports, to disable certain features?

tellthemachines commented:
> I suppose another possibility would be an attribute in block.json for child layout supports, to disable certain features?

Yeah, that's what I was thinking! I guess blocks with no intrinsic width/height are a bit of an edge case so it wouldn't be too cumbersome to do that, but are there any downsides to introducing this new API?

tellthemachines committed:
Fix buggy fill state

andrewserong commented on packages/block-library/src/spacer/edit.js:
```diff
@@ -197,7 +254,51 @@ const SpacerEdit = ( {
+			selfStretch !== 'fit' &&
+			! flexSize
+		) {
+			const newSize =
+				inheritedOrientation === 'horizontal'
+					? getSpacingPresetCssVar( width ) || '72px'
+					: getSpacingPresetCssVar( height ) || '100px';
```
This seems to be working a bit better now, but it looks like there's still a couple of issues when dragging a Spacer block in and out of a flex container. Here's a quick gif:

![2023-04-13 15 18 19](https://user-images.githubusercontent.com/14988353/231661693-5688a696-4840-44a0-8096-0c818b24ff2d.gif)

I'm not too sure what's happening, unfortunately I've found it a little hard to figure out, but my observations so far are:

* The conversion from the spacing preset to the `px` value doesn't appear to populate the Fixed input field
* When moving into the flex container block and out again, somewhere along the way, it seems that the size values become stale. So some of the time the value appears to be inconsistent (like it's remembering the one set in the flex state or vice-versa)

tellthemachines replied:
> The conversion from the spacing preset to the px value doesn't appear to populate the Fixed input field

It's not converting, so the input doesn't understand the value of the preset. Perhaps we should make it possible to set preset sizes in the child layout controls; I'm sure it would be useful outside of this scenario?

> When moving into the flex container block and out again, somewhere along the way, it seems that the size values become stale

Oooh well spotted! We're alternately setting `height` and `flexSize` and not changing either of them when setting the other, so the block attributes end up something like: `{"height":"var:preset|spacing|40","style":{"layout":{"flexSize":"130px","selfStretch":"fixed"}}}`. I'll look into it!

andrewserong replied:
> It's not converting

Ah, I think I'm following along now. I like the idea of adding support for presets for the child layout controls, but I wonder if we'd ultimately want a different scale for size presets than spacing, given that width and height values will often need to be much bigger than spacing between things?

For now, what do you think about potentially using `getCustomValueFromPreset` instead of `getSpacingPresetCssVar` so that when dragging from outside a Row / Stack block into one of them, we switch to using the real pixel value?

tellthemachines replied:
> I wonder if we'd ultimately want a different scale for size presets than spacing, given that width and height values will often need to be much bigger than spacing between things?

Yeah, that's a great point! I'll convert to px for now.

andrewserong commented on packages/block-library/src/spacer/edit.js:
```diff
@@ -197,7 +254,51 @@ const SpacerEdit = ( {
+				setAttributes( {
+					height: '0px',
+				} );
+			}
+		}
+	}, [
+		blockStyle,
+		flexSize,
+		height,
+		inheritedOrientation,
+		isFlexLayout,
+		layout,
+		selfStretch,
+		setAttributes,
+		width,
+	] );
```
Does this mean that the logic here will be called more frequently now? And if so, is that a safe change? From a quick skim it sounds like it's intentional for it to be called more frequently, but I wasn't sure if there's a potential state it could get in where `setAttributes` could result in the effect being run too often 🤔

tellthemachines replied:
Yeah that's a good point, I'm not 100% sure there won't be nefarious side effects. I mainly added the dependencies in because the linter was complaining about it 😅 and it didn't seem to do any harm, but negative perf impacts often don't show immediately.

tellthemachines replied:
Update: I tried removing them and ended up re-adding as the effect wasn't running when a Spacer changed to "fit" or "fill" and the old width/height value was still showing up when dragging outside of the flex container. It doesn't seem to hugely increase the times the effect runs though I might easily be missing some edge case situations.

andrewserong replied:
In practice, I don't think I've encountered any issues with this list of dependencies so far. It might just be something for us to keep an eye on if anyone runs into any issues with the spacer?

andrewserong reviewed (commented):
This is feeling closer to me! I ran into another issue while re-testing, but one thing I did notice is that now that `Fit` is never used as a default value, it doesn't seem to be as offputting that it results in `0` being applied. I think I'd probably lean toward leaving `Fit` in as an option for this PR, and we could look at reducing the available options potentially in a follow-up PR?

tellthemachines committed:
Fix size discrepancies while dragging

tellthemachines committed:
re-add dependencies

andrewserong reviewed (commented):
This is testing really well, it feels very close to being mergeable to me! Just left a comment about whether we should swap from CSS variables to `px` when moving styles from width/height to flexSize.

The other thing I noticed is that when dragging into the flex container, the `setAttributes` call doesn't clear out the `width` or `height` attributes, so in the markup and on the site frontend, it's still outputting the inline style for `width` or `height` respectively. The flex-basis size winds up overriding it, but for tidiness, should we clear those values when the `flexSize` is updated?

Here's some block markup:

```
<!-- wp:spacer {"height":"223px","style":{"layout":{"flexSize":"78px","selfStretch":"fixed"}}} -->
<div style="height:223px" aria-hidden="true" class="wp-block-spacer"></div>
<!-- /wp:spacer -->
```

What do you think? Thanks for all the back and forth — I don't think we need to get it perfect in this PR, so I reckon we could get pretty close and continue to tweak in follow-ups if that winds up being easier.

tellthemachines committed:
Change presets to custom values in flex containers

tellthemachines committed:
actually remove unnecessary attributes

tellthemachines commented:
Thanks for the latest round of testing @andrewserong ! I've updated the PR to change presets to unit values and remove unnecessary attributes when moving between containers.

andrewserong commented on packages/block-editor/README.md:
```diff
@@ -430,6 +430,19 @@ _Returns_
 
 -   `string|null`: A font-size value using clamp().
 
+### getCustomValueFromPreset
```
I totally forgot that we're of course working in a different package when making changes to the Spacer block, so this function needs to be exported now.

Shall we rename it to `getSpacingCustomValueFromPreset` or `getCustomSpacingValueFromPreset` since it's being exported from the block editor package? That way if we have other functions for different kinds of presets the need to be exported in the future, we won't have a collision?

That said, from a quick search I see that the native version of `index.native.js` for the block editor package is already exporting this function as-is here: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-editor/src/components/index.native.js#L79.

So, perhaps another option would be to leave it as-is and if in the future we need a more generic `getCustomValueFromPreset` to be exported that handles other kinds of presets, then we could add a parent function that internally calls the existing spacing one.

What do you think?

tellthemachines replied:
I'd leave it as is for now, and then perhaps if/when we need to handle other types of presets, we can reimplement this function to deal with any preset, so we don't have to change existing usage (we'd probably want to move it out of the spacing utils if we did that of course).

andrewserong commented on packages/block-library/src/spacer/edit.js:
```diff
@@ -191,13 +252,93 @@ const SpacerEdit = ( {
+					...blockStyle,
+					layout: {
+						...layout,
+						flexSize: null,
+						selfStretch: null,
```
Nit: it looks like setting to `null` results in `null` being serialized to the block markup. Should these be `undefined` instead?

E.g. here's block markup after this is updated (it's similar with the `height` and `width` attributes above):

```html
<!-- wp:spacer {"height":"clamp(7rem, 14vw, 11rem)","style":{"layout":{"flexSize":null,"selfStretch":null}}} -->
<div style="height:clamp(7rem, 14vw, 11rem)" aria-hidden="true" class="wp-block-spacer"></div>
<!-- /wp:spacer -->
```

tellthemachines replied:
Oh whoops, I fixed that and then didn't push the change 😳

tellthemachines replied:
Done!

andrewserong reviewed (commented):
This is so very close! Unfortunately, I've run into an odd block validation error and I can't quite work out what's causing it. To reproduce:

1. Add a spacer block outside of a Stack block and set a custom height
2. Drag the block into a Stack block
3. Save the post
4. Reload, and see that there is now a block validation error:

https://user-images.githubusercontent.com/14988353/232675784-6d570e93-9a97-472f-9936-cb6f3f1b069b.mp4

Here's the error message:

```
Content generated by `save` function:

<div style="height:100px" aria-hidden="true" class="wp-block-spacer"></div>

Content retrieved from post body:

<div aria-hidden="true" class="wp-block-spacer"></div>
```

I'm wondering if it has something to do with setting `height` and `width` to `undefined` now? Since the `height` attributes's default value in `block.json` is `100px`, could it be to do with when height is explicitly set to `undefined` versus when it simply has no value and defaults to `100px` when the page is loaded?

I'm not quite sure what's going on there, but one potential fix might be for us to retain setting `width` and `height` to `null` when clearing them out (but leave setting the flex layout values to `undefined` 🤔

tellthemachines committed:
Fix block validation error and remove redundant inline styles

tellthemachines commented:
> I'm wondering if it has something to do with setting height and width to undefined now?

Yeah this is a funny one. The error only goes away when setting an explicit height value (null causes the same error in my testing) so I've changed it to `0px` for now. 
I also removed the extra width and height that were getting added inline to children of flex containers.

Hopefully all is working as expected now 😅

andrewserong commented on packages/block-library/src/spacer/edit.js:
```diff
@@ -191,13 +248,93 @@ const SpacerEdit = ( {
 			setAttributes( {
-				height: '0px',
-				width: '72px',
+				style: {
```
One final nit (and not a blocker), but I just noticed that sometimes the serialized block markup contains an empty `layout` object within `style`. We can use [cleanEmptyObject](https://github.com/WordPress/gutenberg/blob/a3faaef2fdbabac4b17a80f7064ee6b835513b52/packages/block-library/src/utils/clean-empty-object.js#L12) to remove these, e.g. like in the Cover block here: https://github.com/WordPress/gutenberg/blob/7fd28e452957db5228882e978c9ceb4d2be76bb3/packages/block-library/src/cover/transforms.js#L222

It doesn't cause any block validation issues to retain the empty objects in the block markup, though, so I really don't think it's a blocker.

tellthemachines replied:
Funnily enough, running the object passed to `setAttributes` through `cleanEmptyObject` results in the layout attributes not being unset at all, I'm not sure why. I might leave it as is and avoid going down a rabbit hole with this 😅

andrewserong replied:
>  I might leave it as is and avoid going down a rabbit hole with this 😅

Good thinking! I'm in support of the idea of merging this now that it's in a good and working state, and we can look at further tweaks in isolation 👍

andrewserong reviewed (approved):
Okay, I think that's finally gotten to the bottom of the edge cases for me. Thanks so much again for your patience, and for all the back and forth! 🙇 I've left one tiny non-blocking nit, but this is all working well:

✅ Dragging from root/outside Row or Stack to within the Flex block retains the styling correctly
✅ When using presets that have a `px` value, that value is pre-populated in the Fixed input field
✅ Fit and Fill work as expected based on the discussion thus far, with known drawback for `Fit`, but it isn't a default value, so seems fine to retain in the UI for now
✅ For themes using presets that are non-px values, e.g. TT3 that uses `clamp()` rules, the value is still applied and works in the editor and on the site frontend, but the input field is empty. This is to be expected since the child layout controls don't yet support a presets slider or other kind of UI, and I think an empty field is the most graceful fallback we have at the moment 👍
✅ Default values work correctly when inserting a new Spacer block
✅ Spacer block works correctly outside of flex blocks

LGTM! ✨

richtabor commented:
> How would the resize handle be able to work with, say, percentages? (Apologies if I'm missing something obvious.)
> They might not work with percentages - that's probably why the custom Spacer unit control doesn't allow them - but we could conceivably make them work with rem, em, vw and vh. I agree it's not a super high priority though!

Actually, they do work (within a row/stack): 

<img width="1335" alt="CleanShot 2023-04-19 at 18 21 32" src="https://user-images.githubusercontent.com/1813435/233212495-56d59393-4cef-44aa-95d1-9708ad1a5e31.png">
